src/cabal-details.js

```diff
@@ -0,0 +1,365 @@
+      if (!keepUnread) currentChannel.markAsRead()
+      currentChannel.close()
+    }
+    this.chname = channel
+    currentChannel = this.channels[channel]
+    currentChannel.open()
+    this._emitUpdate()
+  }
+
+  closeChannel(channel=this.chname, newChannel) {
+    this.channels[channel].close()
+    // open a new channel after closing `channel`
+    if (newChannel) this.openChannel(newChannel)
+  }
+
+  addStatusMessage(message, channel=this.chname) {
+    this.channels[channel].addMessage(message)
+    this._emitUpdate()
+  }
+
+  getChannels() {
+    return Object.keys(this.channels).sort()
+  }
+
+  getJoinedChannels() {
+    return Object.keys(this.channels).sort()
+      .filter((ch) => this.channels[ch].joined)
+  }
+
+  getJoinedUsers (channel=this.chname) {
+    var details = this.channels[channel]
+    if (!details) return []
+    return details.getMembers()
+  }
+
+  // returns a ChannelDetails object
+  getChannel(channel=this.chname) {
+    return this.channels[channel] 
+  }
+
+  getCurrentChannel() {
+      return this.chname
+  }
+
+  getCurrentChannelDetails() {
+      return this.channels[this.chname]
+  }
+
+  clearVirtualMessages(channel=this.chname) {
+    return this.channels[channel].clearVirtualMessages()
+  }
+
+  getJoinedChannels() {
+    return Object.keys(this.channels).filter(c => this.channels[c].joined).sort()
+  }
+
+  getLocalUser() {
+    return this.user
+  }
+
+  getLocalName() {
+      return this.user.name || this.user.key.slice(0,8)
+  }
+  
+  joinChannel(channel) {
+    var details = this.channels[channel]
+    // we created a channel
+    if (!details) {
+        details = new ChannelDetails(this._cabal, channel)
+        this.channels[channel] = details
+    }
+    // we weren't already in the channel, join
+    if (!details.join()) { 
+      var joinMsg = {
+        type: "channel/join",
+        content: { channel }
+      }
+      // publish a join message to the cabal to signify our presence
+      this._cabal.publish(joinMsg)
+    }
+    // we probably always want to open a joined channel?
+    this.openChannel(channel)
+  }
+
+  leaveChannel(channel) {
+    if (!channel) channel = this.chname
+    if (channel === "!status") return
+    var joined = this.getJoinedChannels()
+    var details = this.channels[channel]
+    if (!details) return
+    var left = details.leave()
+    // we were in the channel, leave
+    if (left) { 
+      var leaveMsg = {
+        type: "channel/leave",
+        content: { channel }
+      }
+      this._cabal.publish(leaveMsg)
+    }
+    var indexOldChannel = joined.indexOf(channel)
+    var newChannel
+    // open up another channel if we left the one we were viewing
+    if (channel === this.chname) {
+      let newIndex = indexOldChannel + 1 
+      if (indexOldChannel >= joined.length) newIndex = 0
+      newChannel = joined[newIndex] || "!status"
+    }
+    this.closeChannel(channel, newChannel)
+  }
+
+  getUsers() {
+    return this.users
```

Comment:
If we return the object instance, any modifications made to it on the outside will affect this object's internal state. Maybe

```suggestion
    return Object.assign({}, this.users)
```